README.md
<a href="https://github.com/open-flash/open-flash">
    <img src="https://raw.githubusercontent.com/open-flash/open-flash/master/logo.png"
    alt="Open Flash logo" title="Open Flash" align="right" width="64" height="64" />
</a>

# SWF Types

[![npm](https://img.shields.io/npm/v/swf-types.svg)](https://www.npmjs.com/package/swf-types)
[![crates.io](https://img.shields.io/crates/v/swf-types.svg)](https://crates.io/crates/swf-types)
[![GitHub repository](https://img.shields.io/badge/Github-open--flash%2Fswf--types-blue.svg)](https://github.com/open-flash/swf-types)
[![Build status](https://img.shields.io/travis/com/open-flash/swf-types/master.svg)](https://travis-ci.com/open-flash/swf-types)

This project specifies an Abstract Syntax Tree (AST) representation of SWF files (SWF movies). It
also includes Rust and Typescript implementations providing type declarations and serialization to
JSON.

Current status: Still a prototype. As a first draft, all the nodes defined in the Flash 10 spec
were implemented in Typescript. We're now revisiting the nodes and progressively adding them
to the documentation while updating the Rust code. The goal of this phase is to align the AST
more with the usage made by Shumway and Flex instead of the grammar. For example the multiple
versions of the "DefineShapeX" grammar rule can be unified in a single `DefineShape` AST node.
What's done in this phase? The general structure (movie, tags), shapes (records, styles), AMV1
actions, partially text support. The main remaining parts are media (audio/video) and AVM2.

The purpose of the AST is to provide a model for the content of SWF file that is a suitable
middle point between the disk representation and the needs of higher-level processing functions.
Beyond processing, the AST is also serializable in unambiguous way to JSON. This allows the
various component manipulating the AST to have a common language.

References:
- [Official SWF 19 spec (mirror)](https://github.com/open-flash/open-flash/blob/master/specs/swf-spec-19.pdf)
- [Shumway SWF file](https://github.com/mozilla/shumway/blob/16451d8836fa85f4b16eeda8b4bda2fa9e2b22b0/src/swf/SWFFile.ts)
- [Flex Movie](https://github.com/apache/flex-sdk/blob/bb7447f5ebe32ce58f85a9e9b2d2ffc48994cf42/modules/swfutils/src/java/flash/swf/Movie.java)
- [SwfMill reference](http://www.swfmill.org/doc/reference.html)

## Documentation

Check the `docs` directory, the root node is called `Movie`.

The structure of the `docs` directory corresponds to the public API of the implementations:
each file represents a type, each directory a namespace.

Example:
```
docs/
├── movie.md
└─┬ tags/
  └── define-shape.md
```
- Typescript:
  ```typescript
  import * as swfTypes from "swf-types";
  const movie: swfTypes.Movie = ...;
  const defineShapeTag: swfTypes.tags.DefineShape = ...;
  ```
- Rust:
  ```rust
  use swf_types;
  fn main() {
    let movie: swf_types::Movie = ...;
    let defineShapeTag: swf_types::tags::DefineShape = ...;
  }
  ```

## Types

A type is a set of values. For example, the type `Boolean` is the set `{true, false}`.
The goal of this library is to describe the type `Movie` which defines the set of valid ASTs.

The documentation uses 4 kinds of types (or "types of types" :P) to build the description of the
AST.

- Primitives are the basic types used to build the other ones (integers, strings, booleans, ...)
- Interfaces are compound types mapping a few keys with their type.
- Enums wrap a few possible sub-types to represent alternatives.
- Collections define containers for same-type values

## Primitive values

The documentation assumes some primitive values such strings, booleans and integers.
The primitives are mostly implementation-dependents. The documentation tries to provide some
parameters to chose the most suitable type available in the language.

## Interface

Most of the AST is defined in term of interfaces. An interface defines a compound type made of
a fixed number of _properties_. Each property has a name (key) and a type.

Example:

> Point in the the sRGB color space with 8-bit color depth.
>
> ```
> interface SRgb8 {
>   r: Uint(8);
>   g: Uint(8);
>   b: Uint(8);
> }
> ```

The type of optional properties is wrapped in `Option()`, for example `Optional(Uint(8))`.

- Typescript: This corresponds to standard TS interfaces. Optional properties use `?`.
- Rust: This corresponds to structs with public fields. Optional properties use `Option<T>`.
- JSON: It corresponds to objects. The keys use `snake_case`. Missing optional properties are not
  emitted.

## Enum

The enums are represented as a tagged union of interfaces.

They are introduced by the keyword `union` followed by the name of the field used as the tag.
This field is added to each variant, its value is the name of the variant in `snake-case`.

Example:

```
union(type) ShapeRecord {
  Edge;
  StyleChange;
}
```

`ShapeRecord` is an enum with two variants: `Edge` and `StyleChange`. They both have an extra
field `type`, with the value `edge` or `style-change` depending on the variant.

## Containers

The main container type is `Vec<T>`: an order list of values of type `T` (can contain any number of
values).

## License

[MIT License](./LICENSE.md)


CHANGELOG.md
# Next

- **[Change]** Match flag order for `DefineFont` and `DefineFontInfo`.

## Rust

- **[Breaking change]** Use bitflags instead of separate `bool` fields in `Bevel`, `ButtonCond`, `ButtonRecord`,
  `ClipEventFlags`, `DefineDynamicText`, `DefineFont`, `DefineFontInfo`, `FileAttributes`, `Glow`, `GradientBevel`,
  `GradientGlow`, `LineStyle` and `MorphLineStyle`.
- **[Breaking change]** Box big variants to reduce size of enums `FillStyle`, `MorphFillStyle`, `MorphShapeRecord`,
  `ShapeRecord` and `Tag`.

# 0.14.0 (2022-05-07)

## Rust

- **[Breaking change]** Require Rust `1.60.0`.
- **[Breaking change]** Rename the feature `use-serde` to `serde`.
- **[Fix]** Update dependencies.

# Typescript

- **[Breaking change]** Update to `kryo@0.14`.
- **[Breaking change]** Compile to `.mjs`.
- **[Fix]** Update dependencies.
- **[Internal]** Use Yarn's Plug'n'Play linker.

# 0.13.0 (2021-07-23)

## Rust

- **[Fix]** Update dependencies.
- **[Fix]** Remove all `unsafe` uses.

## Typescript

- **[Breaking change]** Update to `kryo@0.13`.
- **[Breaking change]** Drop `lib` prefix and `.js` extension from deep-imports.
- **[Fix]** Update dependencies.

# 0.12.0 (2020-09-05)

- **[Breaking change]** Use `PascalCase` for enums and types.
- **[Breaking change]** Require `colorTransform` in `ButtonRecord` ([#85](https://github.com/open-flash/swf-types/issues/85)).
- **[Breaking change]** Rename `DefineButton.characters` to `records` ([#86](https://github.com/open-flash/swf-types/issues/86)).

## Rust

- **[Fix]** Update dependencies.

## Typescript

- **[Breaking change]** Update to native ESM.
- **[Internal]** Switch from `tslint` to `eslint`.

# 0.11.0 (2020-02-05)

- **[Breaking change]** Update `DoAbc` to allow optional flags and name.
- **[Breaking change]** Rename custom image media type to be closer to the corresponding tag names.

## Rust

- **[Internal]** Add clippy support ([#83](https://github.com/open-flash/swf-types/issues/83)).

# 0.10.0 (2019-11-21)

- **[Breaking change]** Rename from `swf-tree`  to `swf-types`.
- **[Breaking change]** Replace `Unknown` tag with `RawBody`.
- **[Feature]** Add `Raw` tag.

## Rust

- **[Feature]** Make `serde` optional (enabled by default) ([#73](https://github.com/open-flash/swf-types/issues/73)).
- **[Fix]** Expose `float_is` module, to allow bit-pattern float equality.
- **[Fix]** Update dependencies.

## Typescript

- **[Fix]** Fix pre-release npm tag.

# 0.9.0 (2019-10-17)

- **[Breaking change]** Change field order in `DefineMorphShape`.
- **[Feature]** Define `DefineButtonSound` tag.
- **[Feature]** Define `DefineButtonColorTransform` tag.
- **[Feature]** Define `DefineVideoStream` tag.
- **[Feature]** Define `EnablePostscript` tag.
- **[Feature]** Define `VideoFrame` tag.
- **[Feature]** Define `SetTabIndex` tag.
- **[Feature]** Define `DefineScalingGrid` tag.
- **[Feature]** Add `image/x-ajpegd` image type to support JPEG with deblocking.
- **[Fix]** Fix `DefineCffFont` definition: reorder fields and ensure data is optional ([#40](https://github.com/open-flash/swf-types/issues/40)).

## Rust

- **[Fix]** Fix `Telemetry` serialization.

# 0.8.0 (2019-07-08)

- **[Breaking change]** Require `language` to be defined in `DefineFontInfo`.
- **[Breaking change]** Reorder `FileAttributes` fields.
- **[Breaking change]** Remove `DefinePartialFont` tag (replaced by `DefineGlyphFont`).
- **[Internal]** Update `README.md`.

## Rust

- **[Breaking change]** Use `[f32; 20]` in `ColorMatrix` filter ([#25](https://github.com/open-flash/swf-types/issues/25)).
- **[Fix]** Fix `Is` implementation for vectors (used to compare floats by bit pattern).
- **[Internal]** Add `rustfmt` support.

# 0.7.2 (2019-07-05)

- **[Feature]** Define `DefineGlyphFont` tag.

# 0.7.1 (2019-07-05)

- **[Feature]** Define `Protect` tag.

### Typescript

- **[Fix]** Update dependencies.

# 0.7.0 (2019-05-21)

- **[Breaking change]** Add `em_square_size` field to `DefineFont` ([#42](https://github.com/open-flash/swf-types/issues/42)).
- **[Breaking change]** Require `align` to be defined in `DefineDynamicText` ([#41](https://github.com/open-flash/swf-types/issues/41)).
- **[Fix]** Update `ButtonCond` field order.
- **[Fix]** Update `ButtonRecord` field order.
- **[Internal]** Update test samples.

### Typescript

- **[Internal]** Update build tools.

# 0.6.0 (2019-04-22)

- **[Breaking change]** Update to MIT/Apache2 license.
- **[Feature]** Define `TagHeader`.
- **[Internal]** Refactor test samples organization.
- **[Internal]** Add `CHANGELOG.md`.

### Rust

- **[Fix]** Skip `control_delta` serialization when `None`.
- **[Fix]** Remove binary-only dependencies from published crate.
- **[Fix]** Update to `swf-fixed@0.1.4`.

### Typescript

- **[Breaking change]** Rename `ClipActions` to `ClipAction` ([#26](https://github.com/open-flash/swf-types/issues/26)).

# 0.5.0 (2019-04-13)

- **[Breaking change]** Unify `StraightEdge` and `CurvedEdge` into `Edge` as well as their morph counterparts. ([#22](https://github.com/open-flash/swf-types/issues/22)).

# 0.4.2 (2019-04-05)

### Rust

- **[Feature]** Derive the following traits when applicable: `Copy`, `Clone`, `PartialOrd`, `Ord` and `Hash`.
- **[Feature]** Implement `Default` for matrices.
- **[Fix]** Update to `swf-fixed@0.1.3`.

# 0.4.1 (2019-03-30)

### Rust

- **[Fix]** Export `SoundEnvelope` and `SoundInfo`
- **[Fix]** Update dependencies

# 0.4.0 (2019-03-29)

- **[Feature]** Add definitions for the following tags: `DoAbc`, `ScriptLimits`, `StartSound`, `StartSound2` and `SymbolClass`

### Rust
- **[Feature]** Derive `Copy` trait for simple enums

# 0.3.2 (2019-03-28)

- **[Breaking change]** Refactor sound tags: reorder fields and rename `streamSoundSampleCount` to `streamSampleCount` and `streamSoundCompression` to `streamFormat`.

# 0.2.1 (2019-03-28)

- **[Fix]** Use `Sint32` for `Rect` dimensions


rs/src/shape.rs
#[cfg(feature = "serde")]
use ::serde::{Deserialize, Serialize};
use static_assertions::const_assert;

use crate::fill_styles;
use crate::join_styles;
#[cfg(feature = "serde")]
use crate::serde_buffer::{buffer_to_hex, hex_to_buffer};

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize), serde(rename_all = "PascalCase"))]
#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum CapStyle {
  None,
  Round,
  Square,
}

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct ClipAction {
  pub events: ClipEventFlags,
  #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
  pub key_code: Option<u8>,
  #[cfg_attr(
    feature = "serde",
    serde(serialize_with = "buffer_to_hex", deserialize_with = "hex_to_buffer")
  )]
  pub actions: Vec<u8>,
}

serde_bitflags! {
  pub struct ClipEventFlags: u32 {
    #[serde(name = "key_up")]
    const KEY_UP = 1 << 7;
    #[serde(name = "key_down")]
    const KEY_DOWN = 1 << 6;
    #[serde(name = "mouse_up")]
    const MOUSE_UP = 1 << 5;
    #[serde(name = "mouse_down")]
    const MOUSE_DOWN = 1 << 4;
    #[serde(name = "mouse_move")]
    const MOUSE_MOVE = 1 << 3;
    #[serde(name = "unload")]
    const UNLOAD = 1 << 2;
    #[serde(name = "enter_frame")]
    const ENTER_FRAME = 1 << 1;
    #[serde(name = "load")]
    const LOAD = 1 << 0;
    #[serde(name = "drag_over")]
    const DRAG_OVER = 1 << 15;
    #[serde(name = "roll_out")]
    const ROLL_OUT = 1 << 14;
    #[serde(name = "roll_over")]
    const ROLL_OVER = 1 << 13;
    #[serde(name = "release_outside")]
    const RELEASE_OUTSIDE = 1 << 12;
    #[serde(name = "release")]
    const RELEASE = 1 << 11;
    #[serde(name = "press")]
    const PRESS = 1 << 10;
    #[serde(name = "initialize")]
    const INITIALIZE = 1 << 9;
    #[serde(name = "data")]
    const DATA = 1 << 8;
    #[serde(name = "construct")]
    const CONSTRUCT = 1 << 18;
    #[serde(name = "key_press")]
    const KEY_PRESS = 1 << 17;
    #[serde(name = "drag_out")]
    const DRAG_OUT = 1 << 16;
  }
}

#[cfg_attr(
  feature = "serde",
  derive(Serialize, Deserialize),
  serde(tag = "type", rename_all = "PascalCase")
)]
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum FillStyle {
  Bitmap(Box<fill_styles::Bitmap>),
  FocalGradient(Box<fill_styles::FocalGradient>),
  LinearGradient(Box<fill_styles::LinearGradient>),
  RadialGradient(Box<fill_styles::RadialGradient>),
  Solid(fill_styles::Solid),
}

const_assert!(std::mem::size_of::<FillStyle>() <= 16);

#[cfg_attr(
  feature = "serde",
  derive(Serialize, Deserialize),
  serde(tag = "type", rename_all = "PascalCase")
)]
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum MorphFillStyle {
  Bitmap(Box<fill_styles::MorphBitmap>),
  FocalGradient(Box<fill_styles::MorphFocalGradient>),
  LinearGradient(Box<fill_styles::MorphLinearGradient>),
  RadialGradient(Box<fill_styles::MorphRadialGradient>),
  Solid(fill_styles::MorphSolid),
}

const_assert!(std::mem::size_of::<MorphFillStyle>() <= 16);

#[cfg_attr(
  feature = "serde",
  derive(Serialize, Deserialize),
  serde(tag = "type", rename_all = "PascalCase")
)]
#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum JoinStyle {
  Bevel,
  Miter(join_styles::Miter),
  Round,
}

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct LineStyle {
  pub width: u16,
  pub start_cap: CapStyle,
  pub end_cap: CapStyle,
  pub join: JoinStyle,
  #[serde(flatten)]
  pub flags: StyleFlags,
  pub fill: FillStyle,
}

serdcfg_attr(feature = "serde", serde(flatten))
  pub struct StyleFlags: u16 {
    #[serde(name = "no_h_scale")]
    const NO_H_SCALE = 1 << 2;
    #[serde(name = "no_v_scale")]
    const NO_V_SCALE = 1 << 1;
    #[serde(name = "no_close")]
    const NO_CLOSE = 1 << 10;
    #[serde(name = "pixel_hinting")]
    const PIXEL_HINTING = 1 << 0;
  }
}

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct MorphLineStyle {
  pub width: u16,
  pub morph_width: u16,
  pub start_cap: CapStyle,
  pub end_cap: CapStyle,
  pub join: JoinStyle,
  #[serde(flatten)]
  pub flags: StyleFlags,
  pub fill: MorphFillStyle,
}

#[cfcfg_attr(feature = "serde", serde(flatten)) = "serde", derive(Serialize, Deserialize))]
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Glyph {
  pub records: Vec<ShapeRecord>,
}

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Shape {
  pub initial_styles: ShapeStyles,
  pub records: Vec<ShapeRecord>,
}

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct MorphShape {
  pub initial_styles: MorphShapeStyles,
  pub records: Vec<MorphShapeRecord>,
}

#[cfg_attr(
  feature = "serde",
  derive(Serialize, Deserialize),
  serde(tag = "type", rename_all = "PascalCase")
)]
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum ShapeRecord {
  Edge(shape_records::Edge),
  StyleChange(Box<shape_records::StyleChange>),
}

const_assert!(std::mem::size_of::<ShapeRecord>() <= 24);

#[cfg_attr(
  feature = "serde",
  derive(Serialize, Deserialize),
  serde(tag = "type", rename_all = "PascalCase")
)]
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum MorphShapeRecord {
  Edge(shape_records::MorphEdge),
  StyleChange(Box<shape_records::MorphStyleChange>),
}

const_assert!(std::mem::size_of::<MorphShapeRecord>() <= 48);

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct ShapeStyles {
  pub fill: Vec<FillStyle>,
  pub line: Vec<LineStyle>,
}

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct MorphShapeStyles {
  pub fill: Vec<MorphFillStyle>,
  pub line: Vec<MorphLineStyle>,
}

pub mod shape_records {
  #[cfg(feature = "serde")]
  use ::serde::{Deserialize, Serialize};

  use crate::Vector2D;

  use super::{MorphShapeStyles, ShapeStyles};

  #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
  #[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
  pub struct Edge {
    /// Difference between the edge start and edge end.
    pub delta: Vector2D,

    /// Difference between the edge start and quadratic bezier control point (if
    /// any).
    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
    pub control_delta: Option<Vector2D>,
  }

  #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
  #[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
  pub struct MorphEdge {
    /// Difference between the edge start and edge end in the start-state of the
    /// morph shape.
    pub delta: Vector2D,

    /// Difference between the edge start and edge end in the end-state of the
    /// morph shape.
    pub morph_delta: Vector2D,

    /// Difference between the edge start and quadratic bezier control point (if
    /// any) in the start-state of the morph shape.
    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
    pub control_delta: Option<Vector2D>,

    /// Difference between the edge start and quadratic bezier control point (if
    /// any) in the end-state of the morph shape.
    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
    pub morph_control_delta: Option<Vector2D>,
  }

  #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
  #[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
  pub struct StyleChange {
    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
    pub move_to: Option<Vector2D>,
    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
    pub left_fill: Option<usize>,
    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
    pub right_fill: Option<usize>,
    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
    pub line_style: Option<usize>,
    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
    pub new_styles: Option<ShapeStyles>,
  }

  #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
  #[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
  pub struct MorphStyleChange {
    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
    pub move_to: Option<Vector2D>,
    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
    pub morph_move_to: Option<Vector2D>,
    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
    pub left_fill: Option<usize>,
    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
    pub right_fill: Option<usize>,
    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
    pub line_style: Option<usize>,
    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
    pub new_styles: Option<MorphShapeStyles>,
  }
}
